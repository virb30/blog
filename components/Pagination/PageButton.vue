<template>
    <button type="button" v-bind="$attrs" :disabled="disabled || active"
        class="h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600"
        :class="buttonClasses" @click="$emit('click', $event)">
        <slot></slot>
    </button>
</template>

<script lang="ts" setup>
import { computed } from 'vue';

defineEmits(['click'])

const props = defineProps({
    active: {
        type: Boolean,
        default: false,
    },
    disabled: {
        type: Boolean,
        default: false,
    },
})

const buttonClasses = computed(() => (
    {
        'opacity-40 cursor-not-allowed ': props.disabled,
        'bg-indigo-500 dark:bg-indigo-500 cursor-default text-gray-50': props.active,
        'hover:opacity-80 cursor-pointer': !props.disabled && !props.active,
    }
));
</script>
