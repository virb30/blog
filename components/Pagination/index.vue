<template>
	<div class="flex flex-col items-center my-12">
		<div class="flex">
			<page-button
				type="button"
				:disabled="!pagination.prev"
				@click="prevPage"
			>
				<FontAwesomeIcon :icon="['fas', 'chevron-left']" />
			</page-button>
			<page-button
				v-for="page in pagination.pages"
				:key="page"
				:active="page === pagination.page"
				@click="toPage(page)"
			>
				{{ page }}
			</page-button>
			<page-button
				:disabled="!pagination.next"
				@click="nextPage"
			>
				<FontAwesomeIcon :icon="['fas', 'chevron-right']" />
			</page-button>
		</div>
	</div>
</template>

<script lang="ts">
import Vue from 'vue';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { mapActions, mapState } from 'vuex';
import { Pagination } from '@tryghost/content-api';
import PageButton from './PageButton.vue';

library.add(faChevronLeft, faChevronRight);

interface Methods {
	prevPage(): void
	toPage(page: number): void
	nextPage(): void
	fetchPosts(page: number): void
}

interface Computed {
	pagination: Pagination,
}

export default Vue.extend<unknown, Methods, Computed, unknown>({
	components: { FontAwesomeIcon, PageButton },
	computed: {
		...mapState('Blog', ['pagination']),
	},
	methods: {
		...mapActions('Blog', ['fetchPosts']),
		nextPage () {
			if (this.pagination.next) {
				this.toPage(this.pagination.next);
			}
		},
		prevPage () {
			if (this.pagination.prev) {
				this.toPage(this.pagination.prev);
			}
		},
		toPage (page: number) {
			this.fetchPosts(page);
		},
	},
});
</script>
