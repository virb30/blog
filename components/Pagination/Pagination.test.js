import { createLocalVue, mount } from '@vue/test-utils';
import Pagination from './index.vue';
import PageButton from './PageButton.vue';

const localVue = createLocalVue();

describe('Pagination Component', () => {
	describe('render', () => {
		it('should render childs', () => {
			const wrapper = mount(Pagination, {
				propsData: {
					pagination: {
						page: 1,
						pages: 2,
						next: 2,
						prev: null,
					},
				},
			});

			const buttons = wrapper.findAll('button');

			expect(wrapper.html()).toMatchSnapshot();
			expect(buttons.length).toBe(4);
		});
	});

	describe('navigation', () => {
		it('should navigate to next page', async () => {
			const wrapper = mount(Pagination, {
				localVue,
				propsData: {
					pagination: {
						page: 1,
						pages: 2,
						next: 2,
						prev: null,
					},
				},
			});

			const toPage = jest.spyOn(wrapper.vm, 'toPage');

			const pageButton = wrapper.findAllComponents(PageButton);

			await pageButton.at(3).vm.$emit('click');
			expect(toPage).toHaveBeenCalledWith(2);
		});

		it('should not navigate', async () => {
			const wrapper = mount(Pagination, {
				localVue,
				propsData: {
					pagination: {
						page: 1,
						pages: 1,
						next: null,
						prev: null,
					},
				},
			});

			const toPage = jest.spyOn(wrapper.vm, 'toPage');

			const pageButton = wrapper.findAllComponents(PageButton);

			await pageButton.at(0).vm.$emit('click');
			expect(toPage).not.toHaveBeenCalled();

			await pageButton.at(2).vm.$emit('click');
			expect(toPage).not.toHaveBeenCalled();
		});

		it('should navigate to prev page', async () => {
			const wrapper = mount(Pagination, {
				localVue,
				propsData: {
					pagination: {
						page: 2,
						pages: 2,
						next: null,
						prev: 1,
					},
				},
			});

			const toPage = jest.spyOn(wrapper.vm, 'toPage');

			const pageButton = wrapper.findAllComponents(PageButton);

			await pageButton.at(0).vm.$emit('click');
			expect(toPage).toHaveBeenCalledWith(1);
		});

		it('should navigate to arbitrary page', async () => {
			const wrapper = mount(Pagination, {
				localVue,
				propsData: {
					pagination: {
						page: 1,
						pages: 3,
						next: 2,
						prev: null,
					},
				},
			});

			const toPage = jest.spyOn(wrapper.vm, 'toPage');

			const pageButton = wrapper.findAllComponents(PageButton);

			await pageButton.at(3).vm.$emit('click');
			expect(toPage).toHaveBeenCalledWith(3);
		});
	});
});
