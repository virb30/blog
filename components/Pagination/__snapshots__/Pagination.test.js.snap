// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component should render childs 1`] = `
"<div class=\\"flex flex-col items-center my-12\\">
  <div class=\\"flex\\"><button type=\\"button\\" disabled=\\"disabled\\" class=\\"h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600 opacity-40 cursor-not-allowed \\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"chevron-left\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 320 512\\" class=\\"svg-inline--fa fa-chevron-left fa-w-10\\">
        <path fill=\\"currentColor\\" d=\\"M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z\\" class=\\"\\"></path>
      </svg></button> <button type=\\"button\\" disabled=\\"disabled\\" class=\\"h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600 bg-indigo-500 dark:bg-indigo-500 cursor-default text-gray-50\\">
      1
    </button><button type=\\"button\\" class=\\"h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600 hover:opacity-80 cursor-pointer\\">
      2
    </button> <button type=\\"button\\" class=\\"h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600 hover:opacity-80 cursor-pointer\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"chevron-right\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 320 512\\" class=\\"svg-inline--fa fa-chevron-right fa-w-10\\">
        <path fill=\\"currentColor\\" d=\\"M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z\\" class=\\"\\"></path>
      </svg></button></div>
</div>"
`;
